unblob/handlers/filesystem/yaffs/utils.py

```diff
@@ -0,0 +1,574 @@
+        yield (start_offset, page, spare)
+        page = file.read(config.page_size)
+        spare = file.read(config.spare_size)
+        start_offset = file.tell()
+
+
+def decode_file_size(high: int, low: int) -> int:
+    """File size can be encoded as 64 bits or 32 bits values.
+
+    If upper 32 bits are set, it's a 64 bits integer value.
+    Otherwise it's a 32 bits value. 0xFFFFFFFF means zero.
+    """
+    if high != 0xFFFFFFFF:
+        return (high << 32) | (low & 0xFFFFFFFF)
+    if low != 0xFFFFFFFF:
+        return low
+    return 0
+
+
+def valid_name(name: bytes) -> bool:
+    # a valid name is either full of null bytes, or unicode decodable
+    try:
+        snull(name[:-1]).decode("utf-8")
+    except UnicodeDecodeError:
+        return False
+    else:
+        return True
+
+
+def is_valid_header(header: Instance) -> bool:
+    if not valid_name(header.name[:-3]):
+        return False
+    if header.type > 5:
+        return False
+    if header.sum_no_longer_used != 0xFFFF:
+        return False
+    return True
+
+
+@attr.define
+class YAFFSEntry:
+    object_type: int
+    object_id: int
+    parent_obj_id: int
+    sum_no_longer_used: int
+    name: str
+    alias: str
+    file_size: int
+    start_offset: int
+
+    def get_chunks(self):
+        raise NotImplementedError
+
+    def __lt__(self, other):
+        return self.object_id < other.object_id
+
+    def __gt__(self, other):
+        return self.object_id > other.object_id
+
+    def __eq__(self, other):
+        return self.object_id == other.object_id
+
+    def __str__(self):
+        return f"{self.object_id}: {self.name}"
+
+
+@attr.define
+class YAFFS1Entry(YAFFSEntry):
+    chunks: List[YAFFS1Chunk]
+
+    def get_chunks(self) -> Iterable[YAFFS1Chunk]:
+        """Return a filtered and ordered list of chunks."""
+        # YAFFS1 chunks have a page_status field indicating
+        # whether or not the chunk is still active. We therefore
+        # filter out chunks that are inactive.
+
+        # YAFFS1 chunks have a serial number that is used to track
+        # which chunk takes precedence if two chunks have the same
+        # identifier. This is used in scenarios like power loss
+        # during a copy operation. Whenever we have two chunks with
+        # the same id, we only return the one with the highest serial.
+
+        for _, chunks in itertools.groupby(
+            sorted(
+                # filter out deleted chunks (page_status=0)
+                [chunk for chunk in self.chunks if chunk.page_status != 0x0],
+                key=lambda chunk: chunk.chunk_id,
+            )
+        ):
+            yield max(chunks, key=lambda chunk: chunk.serial)
+
+
+@attr.define
+class YAFFSFileVar:
+    file_size: int
+    stored_size: int
+    shrink_size: int
+    top_level: int
+
+
+@attr.define(kw_only=True)
+class YAFFS2Entry(YAFFSEntry):
+    chksum: int
+    yst_mode: int
+    yst_uid: int
+    yst_gid: int
```

Comment:
why are these v2 specific, when v1 also has mos of these fields?

Reply:
These fields have been generalized into the `YAFFSEntry` class, they're no longer v2 specific.